File: Assets/Scripts/Networking/Protocol/PlayerInputCodec.cs
using System;

namespace TopDownRoguelike.Networking.Protocol
{
    public sealed class PlayerInputPayload
    {
        public PlayerInputPayload(
    float moveX,
    float moveY,
    float aimX,
    float aimY,
    bool fireHeld = false,
    uint dashRequestSequence = 0u)
        {
            MoveX = moveX;
            MoveY = moveY;
            AimX = aimX;
            AimY = aimY;
            FireHeld = fireHeld;
            DashRequestSequence =
                dashRequestSequence;
        }

        public float MoveX { get; }
        public float MoveY { get; }
        public float AimX { get; }
        public float AimY { get; }
        public bool FireHeld { get; }
        public uint DashRequestSequence { get; }
    }

    public static class PlayerInputCodec
    {
        public const int PayloadSize = 24;

        private const int MoveXOffset = 0;
        private const int MoveYOffset = 4;
        private const int AimXOffset = 8;
        private const int AimYOffset = 12;
        private const int FlagsOffset = 16;
        private const int DashRequestSequenceOffset = 20;
        private const uint FireHeldFlag = 1u;
        private const uint KnownFlags = FireHeldFlag;

        public static byte[] Encode(
            PlayerInputPayload input)
        {
            if (input == null)
            {
                throw new ArgumentNullException(
                    nameof(input));
            }

            Validate(input);

            var payload = new byte[PayloadSize];

            WriteNetworkFloat(
                payload,
                MoveXOffset,
                input.MoveX);

            WriteNetworkFloat(
                payload,
                MoveYOffset,
                input.MoveY);

            WriteNetworkFloat(
                payload,
                AimXOffset,
                input.AimX);

            WriteNetworkFloat(
                payload,
                AimYOffset,
                input.AimY);

            uint flags =
                input.FireHeld
                    ? FireHeldFlag
                    : 0u;

            PacketCodec.WriteNetworkUInt32(
                payload,
                FlagsOffset,
                flags);

            PacketCodec.WriteNetworkUInt32(
                payload,
                DashRequestSequenceOffset,
                input.DashRequestSequence);

            return payload;
        }

        public static PlayerInputPayload Decode(
            byte[] payload)
        {
            if (payload == null)
            {
                throw new ArgumentNullException(
                    nameof(payload));
            }

            if (payload.Length != PayloadSize)
            {
                throw new ArgumentException(
                    "Player input payload has an invalid size.",
                    nameof(payload));
            }

            uint flags =
                PacketCodec.ReadNetworkUInt32(
                    payload,
                    FlagsOffset);

            if ((flags & ~KnownFlags) != 0u)
            {
                throw new ArgumentException(
                    "Player input contains unknown flags.",
                    nameof(payload));
            }

            bool fireHeld =
                (flags & FireHeldFlag) != 0u;

            var input = new PlayerInputPayload(
                ReadNetworkFloat(payload, MoveXOffset),
                ReadNetworkFloat(payload, MoveYOffset),
                ReadNetworkFloat(payload, AimXOffset),
                ReadNetworkFloat(payload, AimYOffset),
                fireHeld,
                PacketCodec.ReadNetworkUInt32(
                    payload,
                    DashRequestSequenceOffset));

            Validate(input);

            return input;
        }

        private static void Validate(
            PlayerInputPayload input)
        {
            if (!IsFinite(input.MoveX) ||
                !IsFinite(input.MoveY) ||
                !IsFinite(input.AimX) ||
                !IsFinite(input.AimY))
            {
                throw new ArgumentException(
                    "Player input contains a non-finite value.");
            }

            if (input.MoveX < -1.0f ||
                input.MoveX > 1.0f ||
                input.MoveY < -1.0f ||
                input.MoveY > 1.0f)
            {
                throw new ArgumentException(
                    "Player movement component is outside [-1, 1].");
            }

            float movementMagnitudeSquared =
                input.MoveX * input.MoveX +
                input.MoveY * input.MoveY;

            if (movementMagnitudeSquared > 1.0001f)
            {
                throw new ArgumentException(
                    "Player movement magnitude exceeds one.");
            }
        }

        private static bool IsFinite(
            float value)
        {
            return !float.IsNaN(value) &&
                !float.IsInfinity(value);
        }

        private static void WriteNetworkFloat(
            byte[] destination,
            int offset,
            float value)
        {
            byte[] localBytes =
                BitConverter.GetBytes(value);

            uint bits =
                BitConverter.ToUInt32(
                    localBytes,
                    0);

            PacketCodec.WriteNetworkUInt32(
                destination,
                offset,
                bits);
        }

        private static float ReadNetworkFloat(
            byte[] source,
            int offset)
        {
            uint bits =
                PacketCodec.ReadNetworkUInt32(
                    source,
                    offset);

            byte[] localBytes =
                BitConverter.GetBytes(bits);

            return BitConverter.ToSingle(
                localBytes,
                0);
        }
    }
}

File: NetworkServer/src/protocol/PlayerInputCodec.cpp
#include "protocol/PlayerInputCodec.h"

#include "protocol/NetworkByteOrder.h"

#include <cmath>
#include <cstring>
#include <stdexcept>

namespace
{
    using tdr::protocol::PlayerInputPayload;

    constexpr std::uint32_t kFireHeldFlag = 1U;
    constexpr std::uint32_t kKnownFlags = kFireHeldFlag;

    static_assert(
        sizeof(float) == sizeof(std::uint32_t),
        "Player synchronization requires 32-bit floats."
        );

    void ValidateInput(
        const PlayerInputPayload& input
    )
    {
        if (!std::isfinite(input.moveX) ||
            !std::isfinite(input.moveY) ||
            !std::isfinite(input.aimX) ||
            !std::isfinite(input.aimY))
        {
            throw std::invalid_argument(
                "Player input contains a non-finite value."
            );
        }

        if (input.moveX < -1.0F ||
            input.moveX > 1.0F ||
            input.moveY < -1.0F ||
            input.moveY > 1.0F)
        {
            throw std::invalid_argument(
                "Player movement component is outside [-1, 1]."
            );
        }

        const float movementMagnitudeSquared =
            input.moveX * input.moveX +
            input.moveY * input.moveY;

        if (movementMagnitudeSquared > 1.0001F)
        {
            throw std::invalid_argument(
                "Player movement magnitude exceeds one."
            );
        }
    }

    void AppendNetwork32(
        std::vector<std::uint8_t>& output,
        const std::uint32_t value
    )
    {
        const std::uint32_t networkValue =
            tdr::protocol::HostToNetwork32(value);

        const auto* bytes =
            reinterpret_cast<const std::uint8_t*>(
                &networkValue
                );

        output.insert(
            output.end(),
            bytes,
            bytes + sizeof(networkValue)
        );
    }

    void AppendNetworkFloat(
        std::vector<std::uint8_t>& output,
        const float value
    )
    {
        std::uint32_t bits = 0U;

        std::memcpy(
            &bits,
            &value,
            sizeof(bits)
        );

        AppendNetwork32(output, bits);
    }

    std::uint32_t ReadNetwork32(
        const std::uint8_t* data,
        std::size_t& offset
    )
    {
        std::uint32_t networkValue = 0U;

        std::memcpy(
            &networkValue,
            data + offset,
            sizeof(networkValue)
        );

        offset += sizeof(networkValue);

        return tdr::protocol::NetworkToHost32(
            networkValue
        );
    }

    float ReadNetworkFloat(
        const std::uint8_t* data,
        std::size_t& offset
    )
    {
        const std::uint32_t bits =
            ReadNetwork32(data, offset);

        float value = 0.0F;

        std::memcpy(
            &value,
            &bits,
            sizeof(value)
        );

        return value;
    }
}

namespace tdr::protocol
{
    std::vector<std::uint8_t>
        PlayerInputCodec::Encode(
            const PlayerInputPayload& input
        )
    {
        ValidateInput(input);

        std::vector<std::uint8_t> encoded;
        encoded.reserve(kPlayerInputPayloadSize);

        AppendNetworkFloat(encoded, input.moveX);
        AppendNetworkFloat(encoded, input.moveY);
        AppendNetworkFloat(encoded, input.aimX);
        AppendNetworkFloat(encoded, input.aimY);
        AppendNetwork32(
            encoded,
            input.flags & kKnownFlags
        );
        AppendNetwork32(
            encoded,
            input.dashRequestSequence
        );

        return encoded;
    }

    PlayerInputPayload
        PlayerInputCodec::Decode(
            const std::uint8_t* const data,
            const std::size_t size
        )
    {
        if (data == nullptr)
        {
            throw std::invalid_argument(
                "Player input payload cannot be null."
            );
        }

        if (size != kPlayerInputPayloadSize)
        {
            throw std::invalid_argument(
                "Player input payload has an invalid size."
            );
        }

        std::size_t offset = 0U;
        PlayerInputPayload input{};

        input.moveX =
            ReadNetworkFloat(data, offset);

        input.moveY =
            ReadNetworkFloat(data, offset);

        input.aimX =
            ReadNetworkFloat(data, offset);

        input.aimY =
            ReadNetworkFloat(data, offset);

        input.flags =
            ReadNetwork32(data, offset);

        input.dashRequestSequence =
            ReadNetwork32(data, offset);

        if ((input.flags & ~kKnownFlags) != 0U)
        {
            throw std::invalid_argument(
                "Player input contains unknown flags."
            );
        }

        ValidateInput(input);

        return input;
    }
}